### ISSUE TITLE
<jemalloc>: Error in munmap(): Cannot allocate memory,errorno=12

### ISSUE THREAD

**Issue by gustars** [jemalloc/jemalloc#970]
Ubuntu 64-bit server, modify extent mmap.c bool opt_retain = false; program a large number of free memory will appear title error

**Comment by davidtgoldblatt** [comment]
Hi, I'm having some trouble understanding what exactly you're hitting here. Could you explain a little bit more?

**Comment by jsteemann** [comment]
Sorry for commenting here, but I hit the same error today, with jemalloc 4.5.0. 
I traced this up to the point when `munmap` actually returns `ENOMEM`. 

It seems that calling munmap can increase the number of mappings, at least when a region is partially unmapped. The *munmap* manpage of my OS has the following to say about munmap returning ENOMEM:

>     ENOMEM The  process's  maximum  number  of  mappings  would  have  been
>              exceeded.  This error can also occur for munmap(2), when  unmap‐
>              ping  a  region in the middle of an existing mapping, since this
>              results in two smaller mappings on either  side  of  the  region
>              being unmapped.

Our version of jemalloc was configured with `--enable-munmap`, and at some point the partial unmaps caused the application to hit the maximum number of mappings (as determined by the value in /proc/sys/vm/max_map_count). Probably the same can be achieved even without `--enable-munmap` when the `retain` option is set to `false`.

What would be interesting is if there is a good workaround for this, i.e. how to make jemalloc return memory to the OS but still avoid the trouble of munmap causing the number of mappings to increase?

**Comment by davidtgoldblatt** [comment]
AFAIK, there is not; this is not a good option to tweak on Linux.

**Comment by davidtgoldblatt** [comment]
(see e.g. the documentation for opt.retain in the man page).

**Comment by interwq** [comment]
@jsteemann : any particular reason you guys use munmap instead of [retain](http://jemalloc.net/jemalloc.3.html#opt.retain) (which is the default)?

**Comment by jsteemann** [comment]
@interwq : at least none that I would know about. 
I saw `--enable-munmap` has been removed in 5.x, so we'll move to `retain` soon.

**Comment by gustars** [comment]
Version: 5.0.1-0-g896ed3a8b3f41998d4fb4d625d30ac63ef2d51fb
Assertions disabled
config.malloc_conf: ""
Run-time option settings:
  opt.abort: false
  opt.abort_conf: false
  opt.retain: true
  opt.dss: "secondary"
  opt.narenas: 32
  opt.percpu_arena: "disabled"
  opt.background_thread: false (background_thread: false)
  opt.dirty_decay_ms: 10000 (arenas.dirty_decay_ms: 10000)
  opt.muzzy_decay_ms: 10000 (arenas.muzzy_decay_ms: 10000)
  opt.junk: "false"
  opt.zero: false
  opt.tcache: true
  opt.lg_tcache_max: 15
  opt.stats_print: false
Arenas: 32
Quantum size: 16
Page size: 4096
Maximum thread-cached size class: 32768
Allocated: 9353024000, active: 10370850816, metadata: 409517576, resident: 10789498880, mapped: 10946904064, retained: 9040003072

**Comment by gustars** [comment]
top - 06:38:18 up 123 days, 22:44,  2 users,  load average: 0.10, 0.10, 0.09
Tasks: 218 total,   1 running, 217 sleeping,   0 stopped,   0 zombie
Cpu(s):  3.5%us,  0.3%sy,  0.0%ni, 96.2%id,  0.0%wa,  0.0%hi,  0.0%si,  0.0%st
Mem:  32880592k total, 27018688k used,  5861904k free,   186640k buffers
Swap:        0k total,        0k used,        0k free,  9525424k cached

  PID USER      PR  NI  VIRT  RES  SHR S %CPU %MEM    TIME+  COMMAND                                                                                                                                                                                                     
22535 rok       20   0 19.0g  15g 2876 S 28.6 49.6   6767:08 skynet

**Comment by gustars** [comment]
sorry,my english is poll。I was in trouble with memory recycle。I suspect that the resident memory is not returned to the system，so i modify opt_retain = false and get this error.

**Comment by gustars** [comment]
Most of the game server process takes up twice as much memory as the memory allocated，so i need help。

**Comment by jsteemann** [comment]
AFAIK (others to correct me if this is wrong) the `opt_retain = false` may cause a situation in which `munmap` can fail with `ENOMEM` errors, which will make the memory allocator return NULL for memory requests even if there is still plenty of free memory available. 

This situation will happen when due to many partial unmaps you reach the maximum number of mappings as dictated by the value in `/proc/sys/vm/max_map_count`.
It may help to set `opt_retain` back to `true` again and enable the background threads (i.e. by setting `background_thread` to `true`). If you do try this, I would be interested in getting to know if this helped or not.

**Comment by gustars** [comment]
Jemalloc Allocated not equal top MEM，If the physical memory is not enough, will use the exchange area？or use the cache or jemalloc？If you use the swap area will cause the program to crash。Looking forward for your reply, thank you。

**Comment by jsteemann** [comment]
@gustars : I was referring to your original problem "Error in munmap(): Cannot allocate memory". I think it may happen if you set retain to `false`, but will not happen when retain is set to `true`. So my suggestion is to set it to `true` again and enable the background threads for purging.
If you at some point run out of physical memory then there is not much you can do except enabling swap or handling any OOM gracefully in your application. This of course requires your OS to be configured in a way that it won't make the OOM killer kill your application first.

**Comment by davidtgoldblatt** [comment]
@jsteemann, yes, that's right: it's easy to hit the 64k limit if creating lots of small ranges.

@gustars, when looking at memory in `top`, are you looking at `VIRT` or `RES`? `VIRT` is expected to be higher with 5.0, but it's also more or less "free" in the sense that it doesn't require physical memory to back it.

Either way, fiddling with opt_retain isn't the right way to go about reducing memory here.

**Comment by gustars** [comment]
Thanks, I will take your advice and keep observing
